examples/provider_upgrade_test.go

```diff
@@ -323,3 +304,67 @@ func providerServer(t *testing.T) pulumirpc.ResourceProviderServer {
 	require.NoError(t, err)
 	return p
 }
+
+type mockPulumiEngine struct {
+	provider                    pulumirpc.ResourceProviderServer
+	lastCheckRequestByURN       map[string]*pulumirpc.CheckRequest
+	verifiedDiffResourceCounter int
+}
+
+func (e *mockPulumiEngine) replayGRPCLog(t *testing.T, jsonLog string) {
+	var entry jsonLogEntry
+	err := json.Unmarshal([]byte(jsonLog), &entry)
+	require.NoError(t, err)
+
+	switch entry.Method {
+	case "/pulumirpc.ResourceProvider/Check":
+		req := unmarshalProto(t, entry.Request, new(pulumirpc.CheckRequest))
+		e.recordCheck(t, req)
+	case "/pulumirpc.ResourceProvider/Diff":
```

Comment:
This makes sense to me, but I had to read through it a couple of times to get it to click.  It would be great to have a comment here that explains what the algorithm is and why it needs to be so complicated. As I understand it, we:
1. Are trying to replay a diff, but
2. the req extracted from `v5(diff, req)` is invalid since `req` is v5 specific.
3. We know that `req = v5(check, rawReq)`, and we have the entire RPC logs, so
4. we retrieve `rawReq` and simulate from there.
5. We use the logs we have to run `reqV6 = v6(check, rawReq)` and then call `v6(diff, reqV6)`, verifying the output of that call.

Reply:
Yeah, exactly. IN case of rds.Instance, this v6(check, ..) corrects "name" to discard the deprecated field; without this the test fails because "name" is then exposed to Diff.